test/unit/source/source_cache.test.js

```diff
@@ -1772,3 +1772,69 @@ test('SourceCache sets max cache size correctly', (t) => {
 
     t.end();
 });
+
+test('SourceCache loads tiles recursively', (t) => {
+    function testSourceCache(name, callback) {
+        t.test(name, (t) => {
+            const transform = new Transform();
+            transform.resize(511, 511);
+            transform.zoom = 14;
+
+            const {sourceCache, eventedParent} = createSourceCache({
+                loadTile (tile, callback) {
+                    if (tile.tileID.canonical.z > 12) {
+                        setTimeout(() => callback({status: 404}), 0);
+                    } else {
+                        tile.state = 'loaded';
+                        callback(null);
+                    }
+                }
+            });
+
+            if (callback) callback(sourceCache);
+
+            eventedParent.on('data', (e) => {
+                if (e.dataType === 'source' && e.sourceDataType === 'metadata') {
+                    sourceCache.update(transform, undefined, sourceCache.usedForTerrain);
```

Comment:
undefined for tileSize is a bit problematic as it doesn't verify real case:

https://github.com/mapbox/mapbox-gl-js/blob/941ba0fdb8d1b945dcd7040c129bff559ed5a5e5/src/terrain/terrain.js#L282-L283

I'd suggest to use transform.zoom = 16, tileSize of `4 * 512` and DEM source maxZoom of 10.

Reply:
I've created a separate test for the terrain https://github.com/mapbox/mapbox-gl-js/blob/75650931288740d57569e13b42b06ff8f2e916df/test/unit/terrain/terrain.test.js#L1547-L1605